service/git/service_test.go

```diff
@@ -46,147 +47,305 @@ func (s *ServiceSuite) TearDownSuite() {
+	dr := NewService(&StorerCommitLoader{s.Storer})
 
+	changeRequests := [2]*lookout.ChangesRequest{
+		&lookout.ChangesRequest{
+			Head: &lookout.ReferencePointer{
+				InternalRepositoryURL: "file:///myrepo",
+				ReferenceName:         "notUsedInTestsButValidated",
+				Hash:                  s.Basic.Head.String(),
+			},
+		},
+		&lookout.ChangesRequest{
+			Base: &lookout.ReferencePointer{
+				InternalRepositoryURL: "file:///myrepo",
+				ReferenceName:         "notUsedInTestsButValidated",
+				Hash:                  "918c48b83bd081e863dbe1b80f8998f058cd8294",
+			},
+			Head: &lookout.ReferencePointer{
+				InternalRepositoryURL: "file:///myrepo",
+				ReferenceName:         "notUsedInTestsButValidated",
+				Hash:                  s.Basic.Head.String(),
+			},
+		},
+	}
+
+	expectedTypes := [2]lookout.ChangeScanner{&TreeScanner{}, &DiffTreeScanner{}}
+	testNames := [2]string{"without base - tree scanner", "with base - diff tree scanner"}
+	for i, changeReq := range changeRequests {
+		s.T().Run(testNames[i], func(t *testing.T) {
+			require := require.New(t)
+			resp, err := dr.GetChanges(context.TODO(), changeReq)
+
+			require.NoError(err)
+			require.NotNil(resp)
+			require.IsType(expectedTypes[i], resp)
+		})
+	}
+}
+
+func (s *ServiceSuite) TestGetChangesWithPattern() {
 	dr := NewService(&StorerCommitLoader{s.Storer})
-	resp, err := dr.GetChanges(context.TODO(), &lookout.ChangesRequest{
-		Head: &lookout.ReferencePointer{
-			InternalRepositoryURL: "file:///myrepo",
-			ReferenceName:         "notUsedInTestsButValidated",
-			Hash:                  s.Basic.Head.String(),
+
+	changeRequests := [2]*lookout.ChangesRequest{
+		&lookout.ChangesRequest{
+			Head: &lookout.ReferencePointer{
+				InternalRepositoryURL: "file:///myrepo",
+				ReferenceName:         "notUsedInTestsButValidated",
+				Hash:                  s.Basic.Head.String(),
+			},
+			IncludePattern: "some-pattern",
 		},
-	})
+		&lookout.ChangesRequest{
+			Base: &lookout.ReferencePointer{
+				InternalRepositoryURL: "file:///myrepo",
+				ReferenceName:         "notUsedInTestsButValidated",
+				Hash:                  "918c48b83bd081e863dbe1b80f8998f058cd8294",
+			},
+			Head: &lookout.ReferencePointer{
+				InternalRepositoryURL: "file:///myrepo",
+				ReferenceName:         "notUsedInTestsButValidated",
+				Hash:                  s.Basic.Head.String(),
+			},
+			ExcludePattern: "some-pattern",
+		},
+	}
+
+	expectedTypesWrapped := [2]lookout.ChangeScanner{&TreeScanner{}, &DiffTreeScanner{}}
+	testNames := [2]string{"without base - tree scanner", "with base - diff tree scanner"}
+	for i, changeReq := range changeRequests {
+		s.T().Run(testNames[i], func(t *testing.T) {
+			require := require.New(t)
+			resp, err := dr.GetChanges(context.TODO(), changeReq)
+
+			require.NoError(err)
+			require.NotNil(resp)
+			require.IsType(&lookout.FnChangeScanner{}, resp)
+			require.IsType(expectedTypesWrapped[i], resp.(*lookout.FnChangeScanner).Scanner)
```

Comment:
hm... Could you please explain to me what exactly this test(s) does? I see it checks types. But why do we need to test implementation details (types) instead of results in response?

Reply:
The idea is that okay that `GetChanges` returns a `lookout.ChangeScanner` which is actually an interface, but depending on the scenario we return different types of `lookout.ChangeScanner`. The idea of this test was to ensure that depending on the scenario we're returning the correct `lookout.ChangeScanner`. I don't know how to test that we're correctly returning the expected scanner except for running it, but there other test cases testing the correct behaviour of each scanner.

> instead of results in response?

What do you mean?

Reply:
> What do you mean?
I mean test `TestGetChangesWithPattern` would check that it returns results only that match pattern. 
And `TestTreeChangesDeleteFile` returns or does not return changes when a file is deleted.

I'm not sure it's a good idea to check which struct implements an interface. For example, if I implement pattern scanner as different structure (without using `FnChangeScanner`) but it would still return only files matching a pattern this test will fail. But all logic is correct and works as expected. It makes me question the test.

@carlosms wdyt?

Reply:
What we should ideally test is the list of `lookout.Change` returned by the `lookout.ChangeScanner` returned by `git.Service::GetChanges` because that's the contract signed by `git.Service::GetChanges`. The process of how the list of `lookout.Change` is built, is an implementation detail of `git.Service::GetChanges` that imo should not be tested.

Reply:
I would prefer to test here too. For example, scanner works fine but we don't pass options to it from service correctly so it returns wrong results.

+1 for refactoring of tests. Repeating checks is error-prone. Service&scanner should return the same result for simple cases when service uses only one scanner.

Reply:
Ok, I got the point. And yes it's better to refactor the tests and also test here the output of the returned scanner. Initially I hadn't given that we already have `scanner_test.go` as @carlosms pointed. But it's true what @smacker says:

> scanner works fine but we don't pass options to it from service correctly so it returns wrong results.

Reply:
So final tests will be:
- `GetChangesInvalidReference` - 1 case
- `GetChangesInvalidTree` - 1 case
- `GetChangesNoBase` - 1 case
- `GetChangesWithPattern` - 3 cases inside for IncludePattern/ExcludePattern/both
- `GetChangesWithContent` - 1 case
- `GetChangesWithPatternAndContent`  - 1 case, tests that scanners don't conflict

that should cover all logic of `GetChanges` methods of the service.

Reply:
another approach for the last 3 cases would be: you check not the type as right now but the instance on scanners and check that the fields passed in the service a correctly set in scanners. Then we have tests for scanners already.
But I really think just re-running scanners would be easier to implement and less fragile.